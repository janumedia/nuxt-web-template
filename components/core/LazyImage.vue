<template>
    <div :class="className">
        <div>
            <img :src="placeHolder" :data-src="src" :alt="title" data-pin-nopin="true">
        </div>
    </div>
</template>
<script>
import {registerListener, unRegisterListener, inViewPort, onCSSTransitionEnd} from "@/assets/js/events/events"
import Vue from 'vue';

const loadImage = (img, index) => {
    
    if(!inViewPort(img, 50)) return;
    
    const src = img.getAttribute("data-src");
    const placeHolder = img.getAttribute("src");
    const parentElementStyle = img.parentElement.style;
    parentElementStyle.backgroundImage = `url(${placeHolder})`;
    parentElementStyle.backgroundRepeat = "no-repeat";
    parentElementStyle.backgroundSize = "cover";
    parentElementStyle.backgroundPosition = "center";
    
    img.onload = () => {
        if(img.src == src) {
            onCSSTransitionEnd(img, () => {
                parentElementStyle.backgroundImage = `url(${img.src})`;//"";
                img.className = "done";
                if(component) component.imageLoaded(index);
            });
            img.onCSS;
            img.className = "image-loaded";
        }
    }
    img.onerror = () => {
        if(component) component.imageError(index);
        img.src = placeHolder;//parentElementStyle.backgroundImage.replace(/(url\(|\)|")/g, '');
    }
    img.className = "image-loading";
    img.removeAttribute("data-src");

    const component = registeredComponent.find(item => item.src == src);
    if(component) component.imageReady(index);

    img.src = src;
}

const onScroll = () => {
    images = document.querySelectorAll(`.${className} img[data-src]`);
    [].slice.call(images).map((img, index) => {
        if (img.complete) loadImage(img, index);
        else {
            Vue.nextTick(() => img.onload = () => loadImage(img, index))
        }
    });
}

const initScrollListener = (f) => {
    registeredComponent.push(f);
    if(init) return;
    //https://nuxtjs.org/faq/window-document-undefined/
    if(process.browser) {
        registerListener("scroll", onScroll);
        registerListener("resize", onScroll);
        //some delay in case new components added
        setTimeout(onScroll);
    }
    init = true;
}

const dispose = (f) => {
    let index = registeredComponent.indexOf(f);
    if(index > -1) registeredComponent.splice(index, 1);
    if(registeredComponent.length <= 0) {
        unRegisterListener("scroll", onScroll);
        unRegisterListener("resize", onScroll);
        init = false;
    }
}

let images, className = "lazy-image", init = false, registeredComponent = [];

export default {
    props:{
        src:{
            type: String,
            require: true
        },
        title: {
            type:String
        }, 
        square: {
            type: Boolean
        }
    },
    data() {
        return { 
            className
        }
    },
    className,
    loadImage,
    computed: {
        placeHolder() {
            return `/images/assets/${this.square ? "2x2" : "3x2"}-placeholder.svg`
        }
    },
    methods: {
        imageReady(index) {
            this.$emit('onReady', this, index);
        },
        imageLoaded(index) {
            this.$emit('onLoaded', this, index);
        },
        imageError(index) {
            this.$emit('onError', this, index);
        }
    },
    beforeCreate() {
        initScrollListener(this);
    },
    beforeDestroy() {
        dispose(this);
    },
}
</script>
<style lang="scss" scoped>
@import "@/assets/css/base.scss";
.lazy-image {
    > div {
        padding: 0 !important;
        margin: 0 !important;
    }
}
img {
    width: 100%;
    opacity: 1;
    display: block;
    @include transition(opacity 0.4s);
    &.image-loading {
        opacity: 0;
        visibility: hidden;
    }
    /*&.done {
        visibility: hidden;
    }*/
}
</style>

