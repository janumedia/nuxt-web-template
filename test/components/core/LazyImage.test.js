import { shallowMount } from '@vue/test-utils'
import LazyImage from '@/components/core/LazyImage'
//
global.process.browser = true;
//
jest.useFakeTimers();

const createWrapper = propsData => shallowMount(LazyImage, {
    attachToDocument: true,
    propsData
})

describe('LazyImage.vue', () => {
    let wrapper,
        IMAGE       = 'image.jpg',
        IMAGE_ERROR = 'image-error.jpg',
        props       = { src: IMAGE, title: 'image title' };
        
    beforeAll(() => {
        LazyImage.components = LazyImage.components || {};
        Object.defineProperty(Image.prototype, 'src', {
            set: function(src) {
                this._src = src;
                if(src == IMAGE_ERROR) this.onerror();
                else this.onload();
            },
            get: function() {
                return this._src;
            }
        })
        
    });

    afterEach(() => {
        //reset
        Element.prototype.getBoundingClientRect = jest.fn(() => {
            return {
                width: 0,
                height: 0,
                top: 0,
                left: 0,
                bottom: 0,
                right: 0,
            }
        });

        Object.defineProperty(Image.prototype, 'complete',  {
            value: false
        });

        //destroy and release scrollListener
        wrapper.destroy();

        //reset setTimer
        jest.clearAllTimers();
    })
    
    it('should render LazyImage init state', () => {
        wrapper = createWrapper(props);
        jest.runOnlyPendingTimers();
        expect(wrapper.html()).toMatchSnapshot();
    });
    
    it('should render LazyImage with square size', () => {
        wrapper = createWrapper(props);
        jest.runOnlyPendingTimers();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render LazyImage on ready state', () => {
        Object.defineProperty(Image.prototype, 'complete',  {
            value: true
        })
        
        wrapper = createWrapper(props);
        
        let spy = jest.fn(() => {
            expect(wrapper.html()).toMatchSnapshot();
        })
        
        wrapper.setMethods({imageReady: spy});
        jest.runOnlyPendingTimers();

        expect(spy).toHaveBeenCalledTimes(1);        
    });

    it('should render LazyImage when load complete', () => {
        Object.defineProperty(Image.prototype, 'complete',  {
            value: true
        })
        wrapper = createWrapper(props);
        jest.runOnlyPendingTimers();
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.emitted().onReady).toBeTruthy();
        expect(wrapper.emitted().onLoaded).toBeTruthy();
    });
    
    it('should render LazyImage when load error', () => {
        Object.defineProperty(Image.prototype, 'complete',  {
            value: true
        });
        wrapper = createWrapper({...props, src: IMAGE_ERROR});
        jest.runOnlyPendingTimers();
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.emitted().onError).toBeTruthy();
    });

    it('should stay in init state when outside viewPort', () => {
        //https://github.com/vuejs/vue-test-utils/issues/319#issuecomment-354667621
        //Element.prototype.scrollTo = jest.fn((x, y) => {});
        //https://stackoverflow.com/a/40006663/1578100
        //https://github.com/facebook/react/issues/4606#issuecomment-130106946
        Element.prototype.getBoundingClientRect = jest.fn(() => {
            return {
                width: 120,
                height: 120,
                top: 1200,
                left: 0,
                bottom: 0,
                right: 0,
            }
        })

        Object.defineProperty(Image.prototype, 'complete',  {
            value: true
        })
        
        wrapper = createWrapper(props);
        
        jest.runOnlyPendingTimers();
        //expect(wrapper.html()).toMatchSnapshot();
        //window.scrollTo(0, 1200);
        expect(wrapper.emitted().onReady).toBeFalsy();
        expect(wrapper.emitted().onLoaded).toBeFalsy();
        expect(wrapper.emitted().onError).toBeFalsy();
    });
})