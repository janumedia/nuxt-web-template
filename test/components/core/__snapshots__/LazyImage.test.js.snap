// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyImage.vue should render LazyImage init state 1`] = `
<div class="lazy-image">
  <div>
    <img src="/images/assets/3x2-placeholder.svg" data-src="image.jpg" alt="image title" data-pin-nopin="true">
  </div>
</div>
`;

exports[`LazyImage.vue should render LazyImage on ready state 1`] = `
<div class="lazy-image">
  <div style="background-image: url(/images/assets/3x2-placeholder.svg); background-repeat: no-repeat; background-size: cover; background-position: center;">
    <img src="/images/assets/3x2-placeholder.svg" alt="image title" data-pin-nopin="true" class="image-loading">
  </div>
</div>
`;

exports[`LazyImage.vue should render LazyImage when load complete 1`] = `
<div class="lazy-image">
  <div style="background-image: url(image.jpg); background-repeat: no-repeat; background-size: cover; background-position: center;">
    <img src="/images/assets/3x2-placeholder.svg" alt="image title" data-pin-nopin="true" class="image-loaded">
  </div>
</div>
`;

exports[`LazyImage.vue should render LazyImage when load error 1`] = `
<div class="lazy-image">
  <div style="background-image: url(/images/assets/3x2-placeholder.svg); background-repeat: no-repeat; background-size: cover; background-position: center;">
    <img src="/images/assets/3x2-placeholder.svg" alt="image title" data-pin-nopin="true" class="image-loading">
  </div>
</div>
`;

exports[`LazyImage.vue should render LazyImage with square size 1`] = `
<div class="lazy-image">
  <div>
    <img src="/images/assets/3x2-placeholder.svg" data-src="image.jpg" alt="image title" data-pin-nopin="true">
  </div>
</div>
`;
